### ISSUE TITLE
Readonly support

### ISSUE THREAD

**Issue by brendt** [spatie/data-transfer-object#12]
I'm not sure how we'd go about this one. But properties on value objects should be read only. I don't think `__set` is of any help, as this are public properties..

**Comment by skyrpex** [comment]
Using protected members will allow `__get` and `__set` to work, but disable autocompletion for me (since they aren't accessible).

Another idea is to use protected members along with getter methods 🤷‍♂️.

**Comment by skyrpex** [comment]
Maybe the ValueObject could parse the public getters? It could work with just this:

```php
class MyValueObject extends ValueObject
{
    public function other(): OtherClass
    {
        return $this->attributes->other;
    }

    public function maybeString(): ?string
    {
        return $this->attributes->maybeString;
    }
}

$value = new MyValueObject([
    "other" => new OtherClass,
    "maybeString" => null,
]);

$value->other();
$value->maybeString();
```

**Comment by brendt** [comment]
Autocompletion and simple syntax is a requirement though. If the syntax becomes more verbose, I don't think the package is as useful anymore..

**Comment by skyrpex** [comment]
Yeah, I totally agree. The syntax is more verbose but the package would rely on the standard reflexion class (which is more maintainable) and it's still better than just creating the ValueObjects manually.

**Comment by YagO-o** [comment]
Hello. Sorry for my bad English.

I faced with same "array vs object" problems while developed my php-projects, and i developed similiar library for Dto-objects. More here:
https://github.com/YagO-o/simple-dto

Always i met with 2 problems when used dtos:
1) Object interface
2) Performance

First problem i decided by getters and IDE autocompletion. I decided, that object interface must be immutable by default, cause faced with problems property overriding in many points of application, cause developer just can it. If you provide bad interface initially, people will be use it in its own way.
I choose way where objects has interface and not has property-access, cause it more obvious solution for interface-based development and protect application against property overriding bugs in anywhere.

Unfortunately, in PHP interface problems may be solved via magic + reflection. And this is very expensive operations for performance. If you use > 10000 dto per your application endpoint lifecycle (this is often may need in background console tasks), execution time became disappointing and dto can be bottleneck. And you may cache property reflection for performance. But you cannot cache access to property via magic.

My ideal Dto is:
1) Immutable, have getters and interfaces instead public properties, and array of properties based initialization in constructor.
2) Have primitive or class-type validation.
3) Fast if you use many dto-objects in your application.
4) Don't very complicated to make class for it.

I vote for getters in this library against public readonly properties for better application performance. 

My library not ideal, but i think is more protected against runtime bugs. 

I hope i pushed some ideas for this library. Is good, and i want use it in future, after some performance and interface features.

**Comment by mfn** [comment]
> interfaces

Can you expand a bit what interfaces you're talking about and showing an example how you would use it?

**Comment by YagO-o** [comment]
> > interfaces
> 
> Can you expand a bit what interfaces you're talking about and showing an example how you would use it?

Under "interfaces" i understand public methods of class to property access. What can describe any other interface if you would.

Example
``` php
class DtoClass extends SimpleDto
{
    /** @var integer */
    private $id;

    /** @var string|null */
    private $name;

    /** @var int[] */
    private $idList;

    /** @var TypeClass */
    private $typeClass;

    /** @var float */
    private $rating = 3.5;
    
    /** @var bool */
    private $isAdmin = false;

    /**
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * @return null|string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @return int[]
     */
    public function getIdList(): array
    {
        return $this->idList;
    }

    /**
     * @return TypeClass
     */
    public function getTypeClass(): TypeClass
    {
        return $this->typeClass;
    }

    /**
     * @return float
     */
    public function getRating(): float
    {
        return $this->rating;
    }
    
    /**
     * @return bool
     */
    public function isAdmin(): bool
    {
        return $this->isAdmin;
    }
}

class TypeClass
{
}


$dto = new DtoClass([
    'id' => 5,
    'name' => 'Ivan',
    'idList' => [1,2,3,4,5],
    'typeClass' => new TypeClass(),
    'isAdmin' => true,
]);

$id = $dto->getId(); // 5
$name = $dto->getName(); // 'Ivan'
$idList = $dto->getIdList(); // [1,2,3,4,5]
$rating = $dto->getRating(); // 3.5
$typeClass = $dto->getTypeClass(); // TypeClass object
$isAdmin = $dto->isAdmin(); // true
```

**Comment by mfn** [comment]
And the interface defines all the getters?

**Comment by YagO-o** [comment]
> And the interface defines all the getters?

Often yes. I often use dto with interface implementation, cause dto is often boundaries between classes communication. 

For example
```
class PlanFilterDto extends SimpleDto implements PlanFilterInterface
{
    /** @var int */
    private $client_id;
    /** @var int */
    private $brand_id;
    /** @var int */
    private $year;

    /**
     * @return int
     */
    public function getClientId(): int
    {
        return $this->client_id;
    }

    /**
     * @return int
     */
    public function getBrandId(): int
    {
        return $this->brand_id;
    }

    /**
     * @return int
     */
    public function getYear(): int
    {
        return $this->year;
    }
}


interface PlanFilterInterface
{
    /**
     * @return int
     */
    public function getBrandId(): int;

    /**
     * @return int
     */
    public function getClientId(): int;

    /**
     * @return int
     */
    public function getYear(): int;
}

class PlanService
{
    /**
     * @param PlanFilterInterface $filter
     *
     * @return PlanDto|null
     */
    public function getByFilter(PlanFilterInterface $filter): ?PlanDto
    {
        return $this->planManager->getByFilter($filter);
    }
}
```

**Comment by brendt** [comment]
Let's keep the following two goals of this package in mind:

- The most easy syntax possible, as less boilerplate as necessary.
- Once typed properties in PHP 7.4 arrive, you should be able to easily update existing DTOs.

Everything this package does can be done with simple getters and setters, without the need of a package. This package is heavily inspired by [Rust's structs](https://doc.rust-lang.org/book/second-edition/ch05-01-defining-structs.html) and I want to keep its use as clean as possible.

I really appreciate everyone thinking about how we can solve the problem of immutability, but as far as I know, we simply can't if we want to keep the above two goals.

**Comment by ryancco** [comment]
This might have been mentioned or tried, but we do have class level property comments [per the docblock documentation](http://docs.phpdoc.org/references/phpdoc/tags/property.html) and can fetch these using [ReflectionClass::getDocComment](http://php.net/manual/en/reflectionclass.getdoccomment.php). These together would provide the type-hinting and IDE-intelligence as well as the immutability by being able to specify the properties as private with public getters and setters.

We can also specify these doc comments specifically as [property-read](http://docs.phpdoc.org/references/phpdoc/tags/property-read.html) to further suggest immutability to the IDE and developer.

**Comment by brendt** [comment]
@ryancco it has been mentioned via mail and on Reddit before, but it's good that you asked it here, so that I can provide a good answer. There are three reasons the `@property` approach is no good fit for this library.

## PHP 7.4

We want this library to be easily used when PHP 7.4 arrives. This means that writing this today:

```php
class Post extends DataTransferObject
{
    /** @var string */
    public $title;
}
```

Can be converted to the following in PHP 7.4, without any breaking changes:

```php
class Post extends DataTransferObject
{
    public string $title;
}
```

Using `@property` or `@property-read` now will make this only more of a hassle. Furthermore: if we want to rely on the 7.4 type system, we'd face the same problem about immutability than.

## Consistency

Not all kinds of types will be supported in PHP 7.4, which is why this package will still be relevant than. Take a look at for example "array of", as we have today:

```php
class Post extends DataTransferObject
{
    /** @var string */
    public $title; 

    /** @var \Tag[] */
    public $tags = [];
}
```

Say we'd use the `@property-read` approach today:

```php
/**
 * @propert-read string $title
 * @propert-read \Tag[] $tags
 */
class Post extends DataTransferObject
{
}
```

And now imagine PHP 7.4 coming along:

```php
/**
 * @propert-read \Tag[] $tags
 */
class Post extends DataTransferObject
{
    public string $title;
}
```

This is the worst of the worst in my opinion. Some properties can be declared within the class, some above.

## Coding style

Let's not forget that doc blocks are never the best solution. I'd want this package to not exist, as it would mean PHP had a proper type system. Doc blocks are a necessary evil at the moment, and they'll still in part in PHP 7.4 because of special types not supported natively.

However, the step towards using `@var` or `@property-read` is quiet different. At least with `@var` we still try to keep as much real PHP code as possible.

In summary: `@property-read` is no good solution for this package, but technically it's a 100% valid approach. I think there's no other solution than to wait for PHP to add readonly properties themselves.

**Comment by dereuromark** [comment]
Performance is usually not a problem by the way, if you generate your classes :)
See https://github.com/dereuromark/cakephp-dto
The idea is to generate an optimized version, instead of doing all that magic and reflection on the go.

I plan to make a more generic (not Cake) bound version, and use the current one as wrapper.
Interested in joining in?

First give it a spin, what you think.

**Comment by brendt** [comment]
Generating PHP code to optimize runtime performance is a great idea, but if you're going that way I'd take it one step further. Do you know of [preprocess.io](https://preprocess.io/) ? It's a project by @assertchris. It's my hope that preprocessing will be accepted by the general community, one day. We saw how much of an impact it had on the JavaScript community, and I think it would be a good way of moving the language forward.

However, I don't think we should try to mimic some of this preprocessing behaviour with stuff like generated classes. It's something a standalone package shouldn't be concerned with. 

In summary, I'm not interested in exploring this area of code generation, as I believe there's much more to be gained if the effort shifted to a more standardized way of generating PHP code.

**Comment by dereuromark** [comment]
Would be interesting to see this in more detail at work.
Until then, my approach seems to cover everything already with the current PHP functionality available and can even mimic the exact type-safety/functionality that will come with future versions (PHP7.4+) in PHP7.1 etc. The nice thing about the generations: Easy to re-run based on future changes in the core language. Nothing on project level needs to change to support more things out of the box.

**Comment by brendt** [comment]
Found a way to fix it! See #50
